ofIsCurrentThreadTheMainThread

artificiel opened the pull request:
while #4040 discusses that `ofThread` should not attempt to maintain a specific relationship to main thread as part of it's design, it is nevertheless useful to be able to determine (without running in a debugger) if a piece of code is called from the main thread or not (especially for monitoring & troubleshooting events, callbacks, notifications and other async constructs)

it is trivial to do so, but requires prior knowledge of the main thread id. this PR stores `ofMainLoop::thread_id` as part of `ofMainLoop` construction, and provides a standardized `ofAppRunner` function that compares it to the current thread::id.

tested on macOS and linux — works, with the presumption that the "main thread" does not change over app lifetime. (is it even possible that something moves the execution of ofMainLoop to another thread? I don't think so but if it happens to be the case `ofMainLoop::thread_id` should be updated after the change).

(PS not 100% sure of the function name... difficult to find a balance between terse and self-documenting)

[edit]: not sure why msys2 is failing but seems unrelated?

ofTheo commented:
All looks good on my end. 
I think maybe having a function for ofMainLoop to return the id might be better but apart from that if you can get it to pass the Mingw tests then be great to merge.

artificiel commented:
OK I've added a getter, made the variable private (and also exposed it globally with ofGetMainThreadId() which might be useful by itself). also added the functions to the basic thread example.

For msys2 I'm confused: the problem was previously a socket networking permission thing, and now it's a linking error with nothing specifically marked as an error -- do you have any insight on how to approach this? (I don't have a setup to test first-hand)

dimitre commented:
I'm having the same issue in a PR test (ucrt64) so I think it is unrelated to PR changes. what do you think @oxillo ?
I think we can temporarly disable ucrt64 test to see if it passes other msys2 tests
